### PR TITLE
Convert compiler-picker to typescript 

### PR DESCRIPTION
**Pull request by jeremy-rifkin**
#2981

### CODE REVIEW

File: static/compiler-picker.ts [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        if (this.tomSelect)
+            this.tomSelect.destroy();
+        this.tomSelect = null;
+    }
+
+    initialize(langId: string, compilerId: string) {
+        this.lastLangId = langId;
+        this.lastCompilerId = compilerId;
+
+        this.tomSelect = new TomSelect(this.domNode as any /*TODO*/, {
```
Comment (RubenRBS) [inline static/compiler-picker.ts:77]: How come it does not understand HTMLElement?
https://tom-select.js.org/docs/#glossary says that NodeDefinition accepts a HTMLElement

File: static/compiler-picker.ts [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+            maxOptions: 1000,
+            onChange: (val) => {
+                if (val) {
+                    ga.proxy('send', {
+                        hitType: 'event',
+                        eventCategory: 'SelectCompiler',
+                        eventAction: val,
+                    });
+                    this.onCompilerChange(val);
+                    this.lastCompilerId = val as any; // TODO
```
Comment (RubenRBS) [inline static/compiler-picker.ts:100]: Why is it not a string here as well?

File: static/compiler-picker.ts [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+                    return '<div class="d-flex"><div>' + escape(data.name) + '</div>' +
+                        '<div title="Click to mark or unmark as a favorite" class="ml-auto toggle-fav">' +
+                        '<i class="' + extraClasses + '"></i>' +
+                        '</div>' +
+                        '</div>';
+                },
+            },
+        });
+
+        $(this.tomSelect.dropdown_content).on('click', '.toggle-fav', evt => {
```
Comment (RubenRBS) [inline static/compiler-picker.ts:117]: Afaik we always add parenthesis to arrow functions

File: static/compiler-picker.ts [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+            evt.preventDefault();
+            evt.stopPropagation();
+
+            if(this.tomSelect) {
```
Comment (RubenRBS) [inline static/compiler-picker.ts:121]: Linter is going to complain about this one missing spaces

File: static/compiler-picker.ts [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    }
+
+    onCompilerFavoriteChange(id: number) {
+        if (this.id !== id) {
+            // Rebuild the rest of compiler pickers so they can properly show the new fav status
+            this.update(this.lastLangId, this.lastCompilerId);
+        }
+    }
+
+    getFavorites(): {[compilerId: string]: boolean} {
```
Comment (RubenRBS) [inline static/compiler-picker.ts:184]: Can you move the type to a `type Favourites = ...`?

File: static/compiler-picker.ts [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+import { ga } from './analytics';
+import * as local from './local';
+
+export class CompilerPicker {
+    static /*const*/ favoriteGroupName = '__favorites__';
+    static /*const*/ favoriteStoreKey = 'favCompilerIds';
+    static nextSelectorId = 1;
+    domRoot: JQuery;
+    domNode: HTMLElement;
+    eventHub: any /* typeof hub.createEventHub() */;
```
Comment (RubenRBS) [inline static/compiler-picker.ts:37]: This comment is probably missleading and should be `ReturnType<typeof hub.createEventHub>` but I know this comes from how it appears in all the other files

**Review by RubenRBS — APPROVED** [review]
Great work, Thanks! :)

File: static/compiler-picker.ts (hunk 1 above)
Comment (jeremy-rifkin) [reply static/compiler-picker.ts:77]: Without the `any` cast I get the following error:

```ts
Argument of type 'HTMLElement' is not assignable to parameter of type 'string | TomInput'.
  Type 'HTMLElement' is missing the following properties from type 'TomInput': disabled, required, value, type, checkValidity
```

I am not sure of the cause of the discrepancy here

File: static/compiler-picker.ts (hunk 4 above)
Comment (jeremy-rifkin) [reply static/compiler-picker.ts:121]: haha good catch 😅

File: static/compiler-picker.ts (hunk 3 above)
Comment (jeremy-rifkin) [reply static/compiler-picker.ts:117]: I see other cases in the source where arrow functions don't have parenthesis. Happy to make the change though, would it also be good to add something to the linter for this?

**Comment by jeremy-rifkin** [comment]
Thank you for taking the time to review! I plan on looking into the tom-select issue more tomorrow

File: static/compiler-picker.ts (hunk 1 above)
Comment (jeremy-rifkin) [reply static/compiler-picker.ts]: Ok, just needed to look at it when I wasn't tired :P Needs to specifically be an HTMLSelectElement or HTMLInputElement

File: static/compiler-picker.ts (hunk 3 above)
Comment (mattgodbolt) [reply static/compiler-picker.ts:126]: Yeah I'm lazy when it comes to arrows, myself. Happy to adopt whatever when we move to prettier/etc though!

**Comment by mattgodbolt** [comment]
Screenshot from 2022-03-16 22-19-30

I just merged and while testing in staging I opened the compiler picker, then the language pickers then...it got stuck in this weird state.

I'm sorry - for now I'm going to revert the merge in -- we can work on it. Really sorry!

**Comment by mattgodbolt** [comment]
Hmm - I can't reproduce now...maybe I'll keep this in...sorry for the indecision. I'd rather fail forward -- so...let's so how this goes! It may have always been an issue, too!

**Comment by mattgodbolt** [comment]
All seems OK: I suspect this is a long-standing issue -- this is now live!

**Comment by jeremy-rifkin** [comment]
Thanks Matt! That's a really strange issue, not something I encountered during testing. If we can reproduce I'll hop on fixing it :)

**Pull request merged**